crates/engine/tree/src/engine.rs

```diff
@@ -0,0 +1,214 @@
+    where
+        T: EngineRequestHandler,
+    {
+        Self { handler, incoming_requests, downloader }
+    }
+}
+
+impl<T, S, D> ChainHandler for EngineHandler<T, S, D>
+where
+    T: EngineRequestHandler,
+    S: Stream<Item = T::Request> + Send + Sync + Unpin + 'static,
+    D: BlockDownloader,
+{
+    type Event = T::Event;
+
+    fn on_event(&mut self, event: FromOrchestrator) {
+        // delegate event to the handler
+        self.handler.on_event(event.into());
+    }
+
+    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<HandlerEvent<Self::Event>> {
+        loop {
+            // drain the handler first
+            loop {
+                match self.handler.poll(cx) {
+                    Poll::Ready(ev) => {
+                        match ev {
+                            RequestHandlerEvent::Idle => break,
+                            RequestHandlerEvent::HandlerEvent(ev) => {
+                                return match ev {
+                                    HandlerEvent::Pipeline(target) => {
+                                        // bubble up pipeline request
+                                        self.downloader.on_action(DownloadAction::Clear);
+                                        Poll::Ready(HandlerEvent::Pipeline(target))
+                                    }
+                                    HandlerEvent::Event(ev) => {
+                                        // bubble up the event
+                                        Poll::Ready(HandlerEvent::Event(ev))
+                                    }
+                                }
+                            }
+                            RequestHandlerEvent::Download(req) => {
+                                // delegate download request to the downloader
+                                self.downloader.on_action(DownloadAction::Download(req));
+                            }
+                        }
+                    }
+                    Poll::Pending => break,
+                }
+            }
+
+            // pop the next incoming request
+            if let Poll::Ready(Some(req)) = self.incoming_requests.poll_next_unpin(cx) {
+                // and delegate the request to the handler
+                self.handler.on_event(FromEngine::Request(req));
+                // skip downloading in this iteration to allow the handler to process the request
+                continue
+            }
+
+            // advance the downloader
+            if let Poll::Ready(DownloadOutcome::Blocks(blocks)) = self.downloader.poll(cx) {
+                // delegate the downloaded blocks to the handler
+                self.handler.on_event(FromEngine::DownloadedBlocks(blocks));
+                continue
+            }
+
+            return Poll::Pending
+        }
+    }
+}
+
+/// A type that processes incoming requests (e.g. requests from the consensus layer, engine API)
+pub trait EngineRequestHandler: Send + Sync {
```

Comment:
```suggestion
pub trait EngineRequestHandler: Send + Sync + Stream<Item = RequestHandlerEvent<Self::Event>> {
```
what is the actual problem here with doing this instead?

Reply:
it's not a stream

Reply:
probably it should be though
> The Stream trait is similar to Future but can yield multiple values before completing
https://rust-lang.github.io/async-book/05_streams/01_chapter.html